### PR TITLE
Implementing the /viewbinds command

### PR DESCRIPTION
**Pull request by One-Nub**
Necessary things to complete before merge:
- [X] Autocomplete for arguments
- [X] Pagination for the viewbind prompt
- [ ] Redesigning/improving the viewbind embed

---

This pull request contains the code implementing the viewbind command. This consists of adding autocomplete handling to the codebase, and adding structures that represent a binding.

### CODE REVIEW

File: src/commands/viewbinds.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+            choices.append(hikari.impl.AutocompleteChoiceBuilder(str(bind), str(bind)))
+
+    # Due to discord limitations, only return the first 25 choices.
+    return interaction.build_response(choices[:25])
+
+
+async def viewbinds_button(interaction: hikari.ComponentInteraction):
+    """Handles the pagination buttons for viewbinds. Since the custom_id includes the next page,
+    we only need one method to handle both buttons."""
+    custom_id_data = interaction.custom_id.split(":")
```
Comment (julien777z) [inline src/commands/viewbinds.py:87]: I wrote a handler for this here https://github.com/bloxlink/bloxlink-http/blob/master/src/resources/component_helper.py#L64

Comment (julien777z) [reply src/commands/viewbinds.py:87]: index here starts at 1 btw

Comment (julien777z) [reply src/commands/viewbinds.py:87]: I'd prob just edit the handler so it can return multiple values with one function call.

File: src/commands/viewbinds.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+    author_id = custom_id_data[1]
+    page_number = custom_id_data[2]
+    category = custom_id_data[3]
+    id_filter = custom_id_data[4]
+
+    # 15 minute timeout on prompts. Woo stateless.
```
Comment (julien777z) [inline src/commands/viewbinds.py:94]: wym timeout, each button press gives u a new 15 mins assuming you defer it

File: src/commands/viewbinds.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        embed = await build_page_embed(page)
+        button_row = page["button_row"]
+        await ctx.response.send(embed=embed, components=button_row)
+
+
+async def build_page_embed(page_components) -> hikari.Embed:
+    embed = hikari.Embed()
+    embed.title = "**Bloxlink Role Binds**"
+
+    bot_user = await bloxlink.rest.fetch_my_user()
```
Comment (julien777z) [inline src/commands/viewbinds.py:186]: this http call isn't necessary

File: src/commands/viewbinds.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    avatar_url = bot_user.default_avatar_url if not bot_user.avatar_url else bot_user.avatar_url
+    embed.set_author(name="Powered by Bloxlink", icon=avatar_url)
+    embed.color = RED_COLOR
+    embed.set_footer("Use /bind to make a new bind, or /unbind to delete a bind")
+
+    if not page_components:
+        page_components = "You have no binds that match the options you passed. "
+        "Please use `/bind` to make a new role bind, or try again with different options."
```
Comment (julien777z) [inline src/commands/viewbinds.py:194]: indent the second line

File: src/commands/viewbinds.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+    if page_components is str:
```
Comment (julien777z) [inline src/commands/viewbinds.py:196]: I don't understand what you're trying to do here

Comment (julien777z) [reply src/commands/viewbinds.py:196]: str is a built-in function and will only return true if page_components is str, which it is not

File: src/commands/viewbinds.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+        embed.description = page_components
+    else:
+        if page_components["linked_group"]:
+            embed.add_field("Linked Groups", "\n".join(page_components["linked_group"]))
+
+        if page_components["group_roles"]:
+            rank_map = page_components["group_roles"]
+            for group in rank_map.keys():
+                embed.add_field(f"{(await get_group(group)).name} ({group})", "\n".join(rank_map[group]))
```
Comment (julien777z) [inline src/commands/viewbinds.py:205]: this will raise an exception if the group is not found (such as an invalid id, or if the group was deleted by roblox)

File: src/commands/viewbinds.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        if page_components["badge"]:
+            embed.add_field("Badges", "\n".join(page_components["badge"]))
+
+        if page_components["gamepass"]:
+            embed.add_field("Gamepasses", "\n".join(page_components["gamepass"]))
+
+    return embed
+
+
+async def build_page_components(
```
Comment (julien777z) [inline src/commands/viewbinds.py:219]: this can prob be abstracted into a common utility for other paginating needs

File: src/commands/viewbinds.py [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+        return (
+            "Your given category option was invalid. "
+            "Only `Group`, `Asset`, `Badge`, and `Gamepass` are allowed options."
+        )
+
+    binds = [GuildBind(**bind) for bind in guild_data.binds]
+
+    filtered_binds = filter(lambda b: b.type == category, binds)
+
+    if id_filter and str(id_filter).lower() != "none":
```
Comment (julien777z) [inline src/commands/viewbinds.py:238]: why are you casting to a str and checking? Can you just do `if id_filter`

File: src/resources/binds.py [hunk 9]
```diff
@@ -246,3 +246,115 @@ async def apply_binds(
...
+                return "linked_group"
+            else:
+                return "group_roles"
+        else:
+            return self.type
+
+    async def get_bind_string(self, guild_id: int, include_id=True, group_data=None) -> str:
+        role_string = await role_ids_to_names(guild_id=guild_id, roles=self.roles)
+        remove_role_str = ""
+        if self.removeRoles:
```
Comment (julien777z) [inline src/resources/binds.py:264]: we don't use camelCase in python

File: src/resources/binds.py [hunk 10]
```diff
@@ -246,3 +246,115 @@ async def apply_binds(
...
+    def determine_type(self) -> str:
+        if self.type == "group":
+            if not self.roles or self.roles == "undefined" or self.roles == "null":
+                return "linked_group"
+            else:
+                return "group_roles"
+        else:
+            return self.type
+
+    async def get_bind_string(self, guild_id: int, include_id=True, group_data=None) -> str:
```
Comment (julien777z) [inline src/resources/binds.py:261]: can you just override `__str__` for this?

File: src/resources/binds.py [hunk 11]
```diff
@@ -246,3 +246,115 @@ async def apply_binds(
         embed = hikari.Embed(description="No binds apply to you!")
 
     return embed
+
+
+class GuildBind(BaseGuildBind):
```
Comment (julien777z) [inline src/resources/binds.py:251]: if we're gonna use GuildBind then it should be used throughout the bot

File: src/resources/binds.py [hunk 12]
```diff
@@ -246,3 +246,115 @@ async def apply_binds(
...
+            )
+        else:
+            return "No valid binding type was given. How? No clue."
+
+        if remove_role_str:
+            bind_string_list.append(remove_role_str)
+        return " → ".join(bind_string_list)
+
+
+async def role_ids_to_names(guild_id: int, roles: list) -> str:
```
Comment (julien777z) [inline src/resources/binds.py:339]: why is this in binds.py

Comment (julien777z) [reply src/resources/binds.py:339]: I'd probably define this in bloxlink.py

File: src/resources/binds.py [hunk 13]
```diff
@@ -246,3 +246,115 @@ async def apply_binds(
...
+
+    for role in roles:
+        output_list.append(
+            guild_roles.get(hikari.Snowflake(role)).name
+            if guild_roles.get(hikari.Snowflake(role)) is not None
+            else "(Deleted Role)"
+        )
+
+    if len(output_list) == 0:
+        return ""
```
Comment (julien777z) [inline src/resources/binds.py:356]: this will already return "" if you call `", ".join(output_list)` with an empty list

File: src/resources/commands.py [hunk 14]
```diff
@@ -64,12 +61,23 @@ async def handle_command(interaction:hikari.CommandInteraction):
...
-        options=command_options
+        options=command_options,
     )
 
     await try_command(command.execute(ctx, subcommand_name=subcommand_name), response)
 
 
+async def handle_autocomplete(interaction: hikari.AutocompleteInteraction):
+    # Iterate through commands and find the autocomplete function that corresponds to the slash cmd option name.
+    for command in slash_commands.values():
```
Comment (julien777z) [inline src/resources/commands.py:72]: sir

Comment (julien777z) [reply src/resources/commands.py:72]: this function is O(n^3)

**Review by julien777z — CHANGES_REQUESTED** [review]

File: src/resources/commands.py (hunk 14 above)
Comment (JohnV50) [reply src/resources/commands.py:72]: agreed

Comment (One-Nub) [reply src/resources/commands.py:72]: I primarily based it on the way handle_component does it, can look into improving big o efficiency of it

File: src/resources/binds.py (hunk 10 above)
Comment (One-Nub) [reply src/resources/binds.py:261]: don't believe __str__ can be async? could be wrong, I can look into it

File: src/commands/viewbinds.py (hunk 8 above)
Comment (One-Nub) [reply src/commands/viewbinds.py:238]: for the custom_id, I include id_filter regardless of it is set or not, so it comes through as `None` as a string literal

Could change the custom_id generator to just not add that element to the custom_id if id_filter is none when making the buttons

File: src/commands/viewbinds.py (hunk 5 above)
Comment (One-Nub) [reply src/commands/viewbinds.py:196]: typically, it is not a str, no
line 227, I return a string if the category is invalid - in which case the bot would reply with that message.

File: src/commands/viewbinds.py (hunk 3 above)
Comment (One-Nub) [reply src/commands/viewbinds.py:186]: guess I'll have to look into another way to get the bot user then, since from my scouring of the hikari docs I couldn't find a way to get the avatar of the bot account without making a rest request.

File: src/commands/viewbinds.py (hunk 2 above)
Comment (One-Nub) [reply src/commands/viewbinds.py:94]: with the stateless nature of the handling, I figured it would be logical that after 15 mintues a prompt could be considered "stale" and the user should just make another rather than having the bot continue to edit that message.

could be changed to just not do anything though, up to you

File: src/commands/viewbinds.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+        id_list = [*set([bind.id for bind in filtered_binds])]
+
+        # Filter id list with user input
+        filtered_binds = filter(lambda b: str(b).startswith(user_input), id_list)
+
+        for bind in filtered_binds:
+            choices.append(hikari.impl.AutocompleteChoiceBuilder(str(bind), str(bind)))
+
+    # Due to discord limitations, only return the first 25 choices.
+    return interaction.build_response(choices[:25])
```
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:81]: The first option is a `view all binds` prompt that shows up regardless of the other options listed (except when matching user input). If a user has over 25 I would infer that they either know the IDs or will just choose view all binds

File: src/commands/viewbinds.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+    # Get user input in a guaranteed manner.
+    user_input = ""
+    for option in interaction.options:
+        if option.name == "category":
+            user_input = option.value
+
+    for bind in binds:
+        bind_type = bind.type
+
+        if bind_type == "asset" and not asset and "asset".startswith(user_input):
```
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:31]: it's this way so that way the prompt can show relevant options when a user puts in a single or a few characters (like `a` or `bad`) rather than having the bot show nothing to autocomplete

the set would get rid of the bools, guess I'd still have to iterate over the set to add the choices but it would result in cleaner code, will fix

File: src/commands/viewbinds.py [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+    group = asset = badge = gamepass = False
+
+    choices = []
+    binds = [GuildBind(**bind) for bind in guild_data.binds]
+
+    # Get user input in a guaranteed manner.
+    user_input = ""
+    for option in interaction.options:
+        if option.name == "category":
+            user_input = option.value
```
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:26]: in our case this shouldn't matter too much since by this point (with the fun O(N^3) code) that option should at least exist (since the interaction isn't dispatched to the method except when it matches our choice name)

but I can look into making it a one-liner

File: src/resources/binds.py (hunk 9 above)
Comment (One-Nub) [reply src/resources/binds.py:264]: removeRoles is how it is setup in the database to my knowledge (looking at quickbinds), I can look into changing the variable to snake_case & see if the dataclass will handle it accordingly or if I'll have to do some conversion

File: src/commands/viewbinds.py [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+        binds = [GuildBind(**bind) for bind in guild_data.binds]
+        filtered_binds = filter(lambda b: b.type.lower() == category_option.value.lower(), binds)
+
+        user_input = id_option.value.lower()
+
+        # Remove duplicate ids
+        id_list = [*set([bind.id for bind in filtered_binds])]
+
+        # Filter id list with user input
+        filtered_binds = filter(lambda b: str(b).startswith(user_input), id_list)
```
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:75]: as far as `==` vs `startswith()` goes, it's handling it that way so that way if a user inputs a "3" it'll show the IDs that start with a 3 (ex `35876262` will show)

iteration count can probably be reduced by using sets instead, will look into it

File: src/commands/viewbinds.py [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+async def viewbinds_id_autocomplete(interaction: hikari.AutocompleteInteraction):
+    category_option = None
+    id_option = None
+
+    base_option = hikari.impl.AutocompleteChoiceBuilder("View all your bindings", "View binds")
+    # Always include the option to view all bindings.
+    choices = [base_option]
+
+    for option in interaction.options:
+        if option.name == "category":
```
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:56]: there's only two options at most in this case though? feels unnecessary to me unless we're really striving for that extra loop cycle to not happen, but with the concept of good practice, I'll add it

File: src/resources/commands.py (hunk 14 above)
Comment (One-Nub) [reply src/resources/commands.py:72]: to elaborate, as I mentioned in a comment earlier-on, its doing all the iterations so that way it's pretty much guaranteed to only dispatch to the right method - I'll still look into improving the efficiency tho 

just from thinking about it, `autocomplete_handlers` is a map, so I should just be able to take the focused option name and call it which would bring big o down to O(N^2) I believe
past that would require more thinking, don't think I can really get lower without just forcing one autocomplete method to handle both/all options which is messy imo

File: src/commands/viewbinds.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (One-Nub) [reply src/commands/viewbinds.py:56]: looking at the code more, there's no need to break out of the loop here? both options are needed for the rest of the code, and again with only two variables it's not that big of a deal 

its if/elif, I could make then both just `if` and continue at the end of their respective if block rather than `if/elif` but they'll effectively work the same
```python
for option in interaction.options:
    if option.name == "category":
        ...
        continue
    
    if option.name == "id":
        ...
        continue
```

Default case, if there's no id then id_option is never set, so it just stays as none

**Review by julien777z — CHANGES_REQUESTED** [review]
~~this is not good, probably revert that, the part where it needs to find a match for the autocomplete handler~~ done

File: src/commands/viewbinds.py [hunk 20]
```diff
@@ -207,15 +214,15 @@ async def build_page_components(
         button_row.add_interactive_button(
             hikari.ButtonStyle.SECONDARY,
             f"viewbinds:{author_id}:{page_number - 1}:{category}:{id_filter}",
-            label="<<",
+            label="\u2B9C",
```
Comment (One-Nub) [inline src/commands/viewbinds.py:217]: The unicode text I chose for both buttons doesn't show on android. Should change, might reuse the button emojis from the current /help prompt on the main bot, not sure yet.

**Comment by DARKVEYDERK01** [comment]
Muy bien

File: src/commands/viewbinds.py (hunk 3 above)
Comment (One-Nub) [reply src/commands/viewbinds.py]: As of right now doesn't look like this can be done; I've asked about it in the hikari support server to see if there is some way.

An alternative is to either not include the image in the prompt, or just hardcode the favicon image that the site uses (since that's the only publicly available place I think the bot icon is hosted which would stay constant).

File: src/commands/viewbinds.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+    embed = await paginator.embed
+    components = paginator.components
+
+    message.embeds[0] = embed
+
+    # Handles emojis as expected
+    # await interaction.edit_message(message, embed=embed, components=[components])
+
+    # TODO: Breaks emojis in the reply somehow?
+    await set_components(message, components=[components])
```
Comment (One-Nub) [inline src/commands/viewbinds.py:94]: looks like this is the "emoji bug" that other devs have talked about, where interaction edits use the 'old' permission system while message edits (which is what set_components) use the 'new' one

tldr of differences:
- old: can use any emoji on platform as along as `@everyone` has use external emojis perm
- new: bot needs to share server with emoji, bot needs perms to use external emojis

not a major issue in my case - just results in some nuances with how i'm styling the embed - might need to shy away from emoji use or just ignore the visual annoyances since my local bot won't be in the server(s) where emojis are stored
